proposals/NNNN-hashable-enhancements.md

```diff
@@ -0,0 +1,925 @@
+these. (The last two algorithms have some light cryptographic elements
+so that they provide reasonable level of protection against hash
+collision attacks.  This makes them better-suited for use in
+general-purpose hashed collections like `Set` and `Dictionary`.)
+
+[FNV-1a]: http://www.isthe.com/chongo/tech/comp/fnv/index.html
+[MurmurHash]: https://github.com/aappleby/smhasher
+[CityHash]: https://github.com/google/cityhash
+[SipHash]: https://131002.net/siphash/
+[HighwayHash]: https://github.com/google/highwayhash
+
+Since [SE-0185] required the standard library to implement a
+high-quality universal hash function, it seems like a good idea to
+expose it as public API, so that manual `Hashable` implementations can
+take advantage of it, too.
+
+Universal hash functions work by maintaining some internal state --
+this can be as simple as a single 32/64-bit integer value (for
+e.g. [FNV-1a]), but it is usually much wider than that. For example,
+[SipHash] maintains a state of 256 bits, while [HighwayHash] uses 1024
+bits. 
+
+## <a name="proposed-solution">Proposed solution</a>
+
+We solve `Hashable`'s implementation problems in two parts. First, we
+make the standard library's hash function public. Second, we replace
+`hashValue` with a requirement that is designed specifically to
+eliminate the guesswork from manual `Hashable` implementations.
+
+### <a name="hasher">The `Hasher` struct</a>
+
+We propose to expose the standard library's standard hash function as
+a new, public struct type, called `Hasher`. This new struct captures
+the state of the hash function, and provides the following operations:
+
+1. An initializer to create an empty state. To make hash values less
+   predictable, the standard hash function uses an arbitrary 128-bit
+   seed value, specified as a tuple of two `UInt64` integers.
+   `Hasher` uses a per-execution random seed value by default;
+   however, we also allow programmers to specify their own seed
+   values, primarily to enable per-instance random seeding in their
+   custom hashing collections.
+   
+    ```swift
+    public struct Hasher {
+      public init()         // Use the default per-execution random seed value
+      public init(seed: (UInt64, UInt64)) // Use the specified seed value
+    }
+    ```
+
+2. A set of operations to mix in new bits into the state of the hash
+   function. For reasons of efficiency, these are built around
+   appending integer values to the hasher:
+   
+    ```swift
+    extension Hasher {
+      public mutating func append(bits: Int)
+      public mutating func append(bits: UInt)
+      public mutating func append(bits: Int64)
+      public mutating func append(bits: UInt64)
+      public mutating func append(bits: Int32)
+      public mutating func append(bits: UInt32)
+      // etc.
+    }
+    ```
+
+    (We expect most hashable types will have discrete components; the
+    bits to be hashed will rarely be available as a contiguous
+    byte sequence.)
```

Comment:
Why? I would expect the vast majority of hashable types to use the default derivation, and many (most?) trivial types to be bitwise-hashable, just as many of them are bitwise-comparable.

Reply:
That's true; however, default derivation isn't aware of bitwise-hashability -- it hashes each individual component separately. We currently have no way of signaling to the compiler that a type is safe to be hashed that way, either. Going through an `UnsafeRawBufferPointer` for each atomic component would be slower than just passing integers.

It isn't even clear yet if Swift's integer types will be bitwise-hashable. It seems `Int` can't be, because it's width depends on the platform, and I really don't want `Int` values to hash like `Int64` on one platform, and `Int32` on another. (For consistency, it may also make sense to ensure that, e.g., `(42 as UInt8)` hashes the same way as `(42 as Int64)`, since they can be compared as equal. I hope we'll not have to do that, though.)

Reply:
> It seems Int can't be, because it's width depends on the platform, and I really don't want Int values to hash like Int64 on one platform, and Int32 on another.

On second thought, the proposal already uses native byte order for multi-byte appends, so it's probably okay to make it even more platform-specific. Alternatively, we could standardize on a particular byte order; there is some advantage in ensuring that the byte sequence we feed to the hasher will be the same on all platforms.

Reply:
Why? What does it matter if the hash value can vary execution-by-execution?